src/nonlinear.jl

```diff
@@ -0,0 +1,240 @@
+# for easy of look-up later.
+for (mathoptformat_string, (julia_symbol, num_arguments)) in SUPPORTED_FUNCTIONS
+    FUNCTION_TO_STRING[julia_symbol] = (mathoptformat_string, num_arguments)
+    STRING_TO_FUNCTION[mathoptformat_string] = (julia_symbol, num_arguments)
+end
+
+"""
+    convert_mof_to_expr(node::Object, node_list::Vector{Object})
+
+Convert a MathOptFormat node `node` into a Julia expression given a list of
+MathOptFormat nodes in `node_list`.
+"""
+function convert_mof_to_expr(node::Object, node_list::Vector{Object})
+    head = node["head"]
+    if head == "real"
+        return node["value"]
+    elseif head == "complex"
+        return Complex(node["real"], node["imag"])
+    elseif head == "variable"
+        return Symbol(node["name"])
+    elseif head == "node"
+        return convert_mof_to_expr(node_list[node["index"]], node_list)
+    else
+        if !haskey(STRING_TO_FUNCTION, head)
+            error("Cannot convert MOF to Expr. Unknown function: $(head).")
+        end
+        (julia_symbol, num_arguments) = STRING_TO_FUNCTION[head]
+        validate_arguments(head, length(node["args"]), num_arguments)
+        expr = Expr(:call, julia_symbol)
+        for arg in node["args"]
+            push!(expr.args, convert_mof_to_expr(arg, node_list))
+        end
+        return expr
+    end
+end
+
+"""
+    convert_mof_to_expr(node::Object, node_list::Vector{Object})
+
+Convert a Julia expression into a MathOptFormat representation. Any intermediate
+nodes that are required are appended to `node_list`.
+"""
+function convert_expr_to_mof(expr::Expr, node_list::Vector{Object})
+    if expr.head != :call
+        error("Expected an expression that was a function. Got $(expr).")
+    end
+    function_name = expr.args[1]
+    if !haskey(FUNCTION_TO_STRING, function_name)
+        error("Cannot convert Expr to MOF. Unknown function: $(function_name).")
+    end
+    (mathoptformat_string, num_arguments) = FUNCTION_TO_STRING[function_name]
+    validate_arguments(function_name, length(expr.args) - 1, num_arguments)
+    node = Object(
+        "head" => mathoptformat_string,
+        "args" => Object[]
+    )
+    for arg in @view(expr.args[2:end])
+        push!(node["args"], convert_expr_to_mof(arg, node_list))
+    end
+    push!(node_list, node)
+    return Object("head" => "node", "index" => length(node_list))
+end
+
+# Recursion end for variables.
+function convert_expr_to_mof(sym::Symbol, node_list::Vector{Object})
+    return Object("head" => "variable", "name" => String(sym))
+end
+
+# Recursion end for real constants.
+function convert_expr_to_mof(value::Real, node_list::Vector{Object})
+    return Object("head" => "real", "value" => value)
+end
+
+# Recursion end for complex numbers.
+function convert_expr_to_mof(value::Complex, node_list::Vector{Object})
+    return Object("head" => "complex", "real" => real(value),
+                  "imag" => imag(value))
```

Comment:
cc @ccoffrin

We can even store complex coefficients. I imagine that might be quite nice for power models???

Reply:
Yes!  And would also be of interest to @JulieSliwak / MathProgComplex.jl

Reply:
Definitely!